letsencrypt-auto-source/rebuild_dependencies.py

```diff
@@ -0,0 +1,221 @@
+versions, ready to be used by pip to install the certbot dependencies.
+
+This script is typically used to update the certbot-requirements.txt file of certbot-auto.
+
+To achieve its purpose, this script will start a certbot installation with unpinned dependencies,
+then gather them, on various distributions started as Docker containers.
+
+Usage: letsencrypt-auto-source/rebuild_dependencies new_requirements.txt
+
+NB1: Docker must be installed on the machine running this script.
+NB2: Python library 'hashin' must be installed on the machine running this script.
+"""
+from __future__ import print_function
+import re
+import shutil
+import subprocess
+import tempfile
+import os
+import sys
+import argparse
+
+# The list of docker distributions to test dependencies against with.
+DISTRIBUTION_LIST = [
+    'ubuntu:18.04', 'ubuntu:14.04',
+    'debian:stretch', 'debian:jessie',
+    'centos:7', 'centos:6',
+    'opensuse/leap:15',
+    'fedora:29',
+]
+
+# Theses constraints will be added while gathering dependencies on each distribution.
+# It can be used because a particular version for a package is required for any reason,
+# or to solve a version conflict between two distributions requirements.
+AUTHORITATIVE_CONSTRAINTS = {
+    # Using an older version of mock here prevents regressions of #5276.
+    'mock': '1.3.0',
+    # Too touchy to move to a new version. And will be removed soon
+    # in favor of pure python parser for Apache.
+    'python-augeas': '0.5.0',
+}
+
+
+# ./certbot/letsencrypt-auto-source/rebuild_dependencies.py (2 levels from certbot root path)
+CERTBOT_REPO_PATH = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
+
+# The script will be used to gathering dependencies for a given distribution.
+#   - certbot-auto is used to install relevant OS packages, and setup an initial venv
+#   - then this venv is used to consistently construct an empty new venv
+#   - once pipstraped, this new venv pip install certbot runtime (including apache/nginx),
+#     without pinned dependencies, and in respect with input authoritative requirements
+#   - `certbot plugins` is called to check we have an healthy environment
+#   - finally current set of dependencies is extracted out of the docker using pip freeze
+SCRIPT = """\
+#!/bin/sh
+set -e
+
+cd /tmp/certbot
+letsencrypt-auto-source/letsencrypt-auto --install-only -n
+PYVER=`/opt/eff.org/certbot/venv/bin/python --version 2>&1 | cut -d" " -f 2 | cut -d. -f1,2 | sed 's/\.//'`
+
+/opt/eff.org/certbot/venv/bin/python letsencrypt-auto-source/pieces/create_venv.py /tmp/venv "$PYVER" "1"
+
+/tmp/venv/bin/python letsencrypt-auto-source/pieces/pipstrap.py
+/tmp/venv/bin/pip install -e acme -e . -e certbot-apache -e certbot-nginx -c /tmp/requirements.txt
+/tmp/venv/bin/certbot plugins
+/tmp/venv/bin/pip freeze >> /tmp/workspace/results
+"""
+
+
+def _process_one_distribution(distribution, verbose):
```

Comment:
I could use a docstring here to delimit the responsibilities of this function, especially since it's just called "process". Docstrings on the other functions would be helpful as well.

Reply:
You are write. These functions make a lot of things. I add a docstring for each of them.